lib/stm32/l4/rcc.c

```diff
@@ -369,4 +454,136 @@ void rcc_set_msi_range_standby(uint32_t msi_range)
+ *
+ * @param clksel One of the definitions above
+ */
+void rcc_set_clock48_source(uint32_t clksel)
+{
+	RCC_CCIPR &= ~(RCC_CCIPR_CLK48SEL_MASK << RCC_CCIPR_CLK48SEL_SHIFT);
+	RCC_CCIPR |= (clksel << RCC_CCIPR_CLK48SEL_SHIFT);
+}
+
+void rcc_clock_setup_hsi16(void)
+{
+	/* Enable internal high-speed oscillator */
+	rcc_osc_on(RCC_HSI16);
+	rcc_wait_for_osc_ready(RCC_HSI16);
+
+	/* Select HSI as SYSCLK source */
+	rcc_set_sysclk_source(RCC_CFGR_SW_HSI16);
+
+	/* Enable high performance mode */
+	rcc_periph_clock_enable(RCC_PWR);
+	pwr_set_vos_scale(PWR_SCALE1);
+
+	/* Set prescalers for AHB, ABP1, ABP2 */
+	rcc_set_hpre(RCC_CFGR_HPRE_NODIV);
+	rcc_set_ppre1(RCC_CFGR_PPRE1_NODIV);
+	rcc_set_ppre2(RCC_CFGR_PPRE2_NODIV);
+
+	/* Configure flash wait states */
+	flash_set_ws(FLASH_ACR_LATENCY_0WS);
+
+	/* Set the peripheral clock frequencies used */
+	rcc_ahb_frequency = 16000000;
+	rcc_apb1_frequency = 16000000;
+	rcc_apb2_frequency = 16000000;
+}
+
+void rcc_clock_setup_pll(const struct rcc_clock_scale *clock)
+{
+	/* Enable selected high-speed oscillator */
+	if (clock->pll_source == RCC_PLLCFGR_PLLSRC_HSE) {
+		rcc_osc_on(RCC_HSE);
+		rcc_wait_for_osc_ready(RCC_HSE);
+	} else if (clock->pll_source == RCC_PLLCFGR_PLLSRC_HSI16) {
+		rcc_osc_on(RCC_HSI16);
+		rcc_wait_for_osc_ready(RCC_HSI16);
+	}
+
+	/* Set performance mode */
+	rcc_periph_clock_enable(RCC_PWR);
+	pwr_set_vos_scale(clock->voltage_scale);
+
+	/* Set prescalers for AHB, ABP1, ABP2 */
+	rcc_set_hpre(clock->hpre);
+	rcc_set_ppre1(clock->ppre1);
+	rcc_set_ppre2(clock->ppre2);
+
+	/* Configure PLL */
+	rcc_set_main_pll(clock->pll_source, clock->pllm, clock->plln,
+			 clock->pllp, clock->pllq, clock->pllr);
+
+	/* Enable PLL and wait for it to stabilize */
+	rcc_osc_on(RCC_PLL);
+	rcc_wait_for_osc_ready(RCC_PLL);
+
+	/* Configure flash wait states */
+	flash_set_ws(clock->flash_waitstates);
+
+	/* Select PLL as SYSCLK source and wait for it to be selected */
+	rcc_set_sysclk_source(RCC_CFGR_SW_PLL);
+	rcc_wait_for_sysclk_status(RCC_PLL);
+
+	/* Set the peripheral clock frequencies used */
+	rcc_ahb_frequency = clock->ahb_frequency;
+	rcc_apb1_frequency = clock->apb1_frequency;
+	rcc_apb2_frequency = clock->apb2_frequency;
+}
+
+/** Enable the RTC clock */
+void rcc_enable_rtc_clock(void)
+{
+	RCC_BDCR |= RCC_BDCR_RTCEN;
+}
+
+/** Disable the RTC clock */
+void rcc_disable_rtc_clock(void)
+{
+	RCC_BDCR &= ~RCC_BDCR_RTCEN;
+}
+
+/** Set the source for the RTC clock
+ * @param[in] clk ::rcc_osc. RTC clock source. Only HSE/32, LSE and LSI.
+ */
+void rcc_set_rtc_clock_source(enum rcc_osc clk)
+{
+	RCC_BDCR &= ~(RCC_BDCR_RTCSEL_MASK << RCC_BDCR_RTCSEL_SHIFT);
+
+	switch (clk) {
+	case RCC_HSE:
+		RCC_BDCR |= (RCC_BDCR_RTCSEL_HSEDIV32 << RCC_BDCR_RTCSEL_SHIFT);
+		break;
+	case RCC_LSE:
+		RCC_BDCR |= (RCC_BDCR_RTCSEL_LSE << RCC_BDCR_RTCSEL_SHIFT);
+		break;
+	case RCC_LSI:
+		RCC_BDCR |= (RCC_BDCR_RTCSEL_LSI << RCC_BDCR_RTCSEL_SHIFT);
+		break;
+	default:
+		/* none selected */
```

Comment:
this will actually result in clearing the source if you provide an invalid one? I guess that's ok?

Reply:
i suppose so, there are many functions in libopencm3 like this

Reply:
some, most of them do nothing if you pass in the wrong thing.  Just making the value to |= a variable would fix that. (not really important though)